File: monadic-parser.php
<?php
# -*- coding: utf-8 -*-
#see http://sandersn->com/blog//index->php/2009/07/01/monadic_parsing_in_python_part_3

#todo:
# add exception property to parseresult
# put in ast result class as dom object
# make stateful parsebuffer object via clone
# remove $lexer
# insert namespace
# use list() = instead of parseresult

function ast_result($type, $value){
  return array('type'=>$type, 'vaule'=>$value);
}

class Ast{
  function add_token($type, $value){
  }
  function add_structure($type, $value, $content){
  }
  function dump() {
  }
}
class ParseBuffer{
  function __construct($s){
    $this->s=$s;
    $this->i=0;
  }
  function eof(){
    return ($this->i == strlen($this->s));
  }
  function match($pattern){
    if(preg_match($pattern, $this->s, $matches, 0, $this->i)){
      $this->i += strlen($matches[0]);
      return $matches[0];
    } else {
      return new Exception("ParseBuffer could not match at {$this->i}.");
    }
  }
}

class ParseResult{
  function __construct(ParseBuffer $buffer, $answer){
    $this->buffer = $buffer;
    $this->answer = $answer;
  }
}

abstract class Parser{
  function __construct(Lexer $lexer){
    $this->lexer = $lexer;
  }
  abstract function parse(ParseBuffer $buffer);
  function parsestring($s){
    $result = $this->parse(new ParseBuffer($s));
    if($result->answer instanceof Exception){ 
      throw $result->answer;
    } elseif($result->buffer->eof()) {
      return $result->answer;
    } else {
      throw new Exception("Parser did not reach EOF, only {$result->buffer->i}");
    }
  }
}


class Token extends Parser{
  function __construct($pattern, $type = 'Token'){
    $this->type = $type;
    $this->pattern = $pattern;
  }
  function parse(ParseBuffer $buffer){
    $token = $buffer->match($this->pattern);
    if($token instanceof Exception){
      return new ParseResult($buffer, $token);
    } else {
      return new ParseResult($buffer, ast_result($this->type, $token));
    }
  }
}

class Chain extends Parser{
  function __construct(&$parsers, $type = 'Chain'){
    $this->type = $type;
    assert(is_array($parsers));
    foreach($parsers as $parser){
      assert($parser instanceof Parser);
    }
    $this->parsers = &$parsers;
  }
  function parse(ParseBuffer $buffer){
    $value = array();
    foreach($this->parsers as $parser){
      $result = $parser->parse($buffer);
      if($result->answer instanceof Exception){
        break;
      }
      $buffer = $result->buffer;
      $value[] = $result->answer;
    }
    return new ParseResult($buffer, ast_result($this->type,$value));
  }
}

class Lift extends Parser{
  function __construct($parser1, $f){
    $this->parser1 = $parser1;
    $this->f = $f;
  }
  function parse(ParseBuffer $buffer){
    $result = $this->parser1->parse($buffer);
    if ($result->answer instanceof Exception){
    return new ParseResult($buffer, $result->answer);
    } else {
      return $this->f(a).parse($buffer);
    }
  }
}
class Rtrn extends Parser{
  function __construct($x){
    $this->x = $x;
  }
  function parse(ParseBuffer $buffer){
    return new ParseResult($buffer, $this->x);
  }
}
class Any extends Parser{
  function __construct(&$parsers, $type = 'Any'){
    $this->type = $type;
    assert(is_array($parsers));
    foreach($parsers as $parser){
      assert($parser instanceof Parser);
    }
    $this->parsers = &$parsers;
  }
  function parse(ParseBuffer $buffer){
    foreach($this->parsers as $parser){
      $result = $parser->parse($buffer);
      if (!($result->answer instanceof Exception)){
        break;
      }
    }
    return new ParseResult($result->buffer, ast_result($this->type,$value));
  }
}

File: monadic-parser.py
# -*- coding: utf-8 -*-
#see http://sandersn.com/blog//index.php/2009/07/01/monadic_parsing_in_python_part_3

class Parser():
    def parse(self, s, i):
        raise NotImplementedError()
    def run(self, s):
        (s,i), answer = self.parse(s, 0)
        if i==len(s) and not isinstance(answer, Exception):
            return answer
        else:
            raise answer
def runsexp(s):
    return Bind(Chain(SkipSpace(), Sexp()),
                lambda answer: Chain(SkipSpace(), Return(answer))).run(s)


class Atom(Parser):
    def parse(self, s, i):
        if i >= len(s): return Fail("Ran off end of string").parse(s,i)
        start = i
        while i < len (s) and not s[i].isspace() and not s[i]==')' and not s[i]=='(':
            i += 1
        if i==start:
            return Fail("Atom not found at %d; '%s' found instead." % (i, s[i])).parse(s,i)
        else:
            return (s,i), s[start:i]
class SkipSpace(Parser):
    def parse(self, s, i):
        while i < len(s) and s[i].isspace():
            i += 1
        return (s,i), ()
class Char(Parser):
    def __init__(self, c):
        self.c = c
    def parse(self, s, i):
        if i >= len(s): return Fail("Ran off end of string").parse(s, i)
        if s[i]==self.c:
            return (s,i+1), self.c
        else:
            return Fail("'%s' expected at %d, but '%s' found instead. [%s]" %
                        (self.c,i,s[i],s[max(0,i-5):min(len(s), i+5)])).parse(s, i)

class Chain(Parser):
    def __init__(self, parser1, parser2):
        self.parser1 = parser1
        self.parser2 = parser2
    def parse(self, s, i):
        (s,i), a = self.parser1.parse(s, i)
        if isinstance(a, Exception): return (s,i), a
        else: return self.parser2.parse(s, i)
class Bind(Parser):
    def __init__(self, parser1, f):
        self.parser1 = parser1
        self.f = f
    def parse(self, s, i):
        (s,i), a = self.parser1.parse(s, i)
        if isinstance(a, Exception): return (s,i), a
        else: return self.f(a).parse(s, i)
class Return(Parser):
    def __init__(self, x):
        self.x = x
    def parse(self, s, i):
        return (s,i), self.x
class Fail(Parser):
    def __init__(self, msg):
        self.msg = msg
    def parse(self, s, i):
        return (s,i), Exception(self.msg)
class Or(Parser):
    def __init__(self, *alts):
        self.alts = alts
    def parse(self, s, i):
        for alt in self.alts:
            (snoo,inew), a = alt.parse(s, i)
            if not isinstance(a, Exception):
                return (snoo,inew), a
        return (snoo,inew), a # (s,i), a

